src/Concerns/TransactionManager.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace Jenssegers\Mongodb\Concerns;
+
+use Closure;
+use MongoDB\Driver\Exception\RuntimeException;
+use MongoDB\Driver\Session;
+use function MongoDB\with_transaction;
+
+trait TransactionManager
+{
+    /**
+     * A list of transaction session.
+     * @var Session|null
+     */
+    protected ?Session $session;
+
+    /**
+     * Get the existing session or null.
+     */
+    public function getSession(): ?Session
+    {
+        return $this->session ?? null;
+    }
+
+    /**
+     * Use the existing or create new session and start a transaction in session.
+     *
+     * In version 4.0, MongoDB supports multi-document transactions on replica sets.
+     * In version 4.2, MongoDB introduces distributed transactions, which adds support for multi-document transactions on sharded clusters and incorporates the existing support for multi-document transactions on replica sets.
+     *
+     * @see https://docs.mongodb.com/manual/core/transactions/
+     * @param array $options
+     * @return void
+     */
+    public function beginTransaction(array $options = []): void
+    {
+        $session = $this->getSession();
+
+        if ($session === null) {
+            $session = $this->connection->startSession();
+            $this->session = $session;
+        }
+
+        $session->startTransaction($options);
+    }
+
+    /**
+     * Commit transaction in this session and close this session.
+     * @return void
+     */
+    public function commit(): void
+    {
+        $this->throwExceptionIfTransactionDoesNotStart();
+        $session = $this->getSession();
+
+        $session?->commitTransaction();
+    }
+
+    /**
+     * Rollback transaction in this session and close this session.
+     * @param null $toLevel
+     * @return void
+     */
+    public function rollBack($toLevel = null): void
+    {
+        $this->throwExceptionIfTransactionDoesNotStart();
+        $session = $this->getSession();
+
+        $session?->abortTransaction();
```

Comment:
> What is the point of using the [null-safe operator](https://php.watch/versions/8.0/null-safe-operator) if you've already verified that the session exists with `throwExceptionIfTransactionDoesNotStart()`? I could imagine this being necessary to satisfy code analysis tools since they might not be able to infer that `throwExceptionIfTransactionDoesNotStart()` guaranteed that `getSession()` would never return `null`; however, that's not relevant here.

Not that it is relevant here, but Psalm is able to handle this via the `@psalm-assert` annotation. See the first example [in the docs](https://psalm.dev/docs/annotating_code/adding_assertions/). Annotating the `throwIfNoSessionStarted` like this will tell psalm that `$this->session` is an instance of `Session`:
```php
/** @psalm-assert Session $this->session */
protected function throwIfNoSessionStarted(): void
[...]
```